Source/DataSources/KmlExporter.js

```diff
@@ -0,0 +1,1165 @@
+        }
+
+        StyleCache.prototype.get = function(element) {
+            var ids = this._ids;
+            var key = element.innerHTML;
+            if (defined(ids[key])) {
+                return ids[key];
+            }
+
+            var styleId = 'style-' + (++this._count);
+            element.setAttribute('id', styleId);
+
+            // Store with #
+            styleId = '#' + styleId;
+            ids[key] = styleId;
+            this._styles[key] = element;
+
+            return styleId;
+        };
+
+        StyleCache.prototype.save = function(parentElement) {
+            var styles = this._styles;
+
+            var firstElement = parentElement.childNodes[0];
+            for (var key in styles) {
+                if (styles.hasOwnProperty(key)) {
+                    parentElement.insertBefore(styles[key], firstElement);
+                }
+            }
+        };
+
+        function IdManager() {
+            this._ids = {};
+        }
+
+        IdManager.prototype.get = function(id) {
+            if (!defined(id)) {
+                return this.get(createGuid());
+            }
+
+            var ids = this._ids;
+            if (!defined(ids[id])) {
+                ids[id] = 0;
+                return id;
+            }
+
+            return id.toString() + '-' + (++ids[id]);
+        };
+
+        /**
+         * Exports an EntityCollection as a KML document. Only Point, Billboard, Model, Path, Polygon, Polyline geometries will be exported. If your
+         * document contains a Model, you must specify a modelCallback to handle the conversion as KML doesn't allow for glTF models. There is not
+         * a 1 to 1 mapping of Entity properties to KML Feature properties, so there are some caveats. Entity properties that are time dynamic but
+         * cannot be dynamic in KML are exported with their values at options.time or the beginning of the EntityCollection if not specified. For time-dynamic
+         * properties that are supported in KML, we use the samples if it is a SampledProperty otherwise we sample the value using the options.sampleDuration.
+         * Point, Billboard, Model and Path geometries with time-dynamic positions will be exported as gx:Track Features. Not all Materials are representable
+         * in KML, so for more advanced Materials just the primary color is used. Canvas objects are exported as png images.
+         *
+         * @alias KmlExporter
+         * @constructor
+         *
+         * @param {EntityCollection} entities The EntityCollection to export as KML
+         * @param {Object} options An object with the following properties:
+         * @param {Function} [options.ellipsoid=Ellipsoid.WGS84] The ellipsoid for the output file
+         * @param {Function} [options.modelCallback] A callback that will be called with a ModelGraphics instance and should return the URI to use in the KML. This will throw a RuntimeError if there is a model and this is undefined.
```

Comment:
You can actually define the specific callback as a type instead of just `Function` so its crystal clear what parameters are expected.  Search for `TimeInterval~MergeCallback` as an example.

Reply:
Yep. I think that's what @OmarShehata was referring to and not textureCallback. I'll add the callback doc and that should keep things straight.